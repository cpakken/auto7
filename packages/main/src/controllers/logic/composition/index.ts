export * from "./connections-map"
export * from "./paths"
export * from "./blocks"
export * from "./composition"
