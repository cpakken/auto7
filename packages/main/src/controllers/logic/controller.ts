import { getEnv, getRoot, IAnyStateTreeNode, Instance, SnapshotIn, SnapshotOut, types } from "mobx-state-tree"
import { LogicComposedStore, LogicNativeStore } from "."

export interface ILogicControllerSnapshotIn extends SnapshotIn<typeof LogicController> {}
export interface ILogicControllerSnapshot extends SnapshotOut<typeof LogicController> {}
export interface ILogicController extends Instance<typeof LogicController> {}
export const LogicController = types.model({
  composed: types.optional(LogicComposedStore, {}),
  native: types.optional(LogicNativeStore, {}),
})

export function getLogicController(instance: IAnyStateTreeNode): ILogicController {
  return getEnv(instance).logicController || getRoot(instance)
}
