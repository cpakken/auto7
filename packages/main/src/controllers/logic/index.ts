export * from "./info"
export * from "./composition"
export * from "./stores"
export * from "./controller"
