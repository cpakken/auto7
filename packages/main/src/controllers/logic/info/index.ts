export * from "./interface"
export * from "./info"