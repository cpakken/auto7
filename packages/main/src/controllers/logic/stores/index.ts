export * from "./native-store"
export * from "./composed-store"
export * from "./LogicNative"
export * from "./LogicComposed"
