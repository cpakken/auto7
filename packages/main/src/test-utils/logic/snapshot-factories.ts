import { LogicNativeSnapshotIn, LogicComposedSnapshotIn, ILogicComposedStoreSnapshotIn } from "@main/controllers/logic"

export function createLogicComposedStoreSnapshot(
  snapshots: LogicComposedSnapshotIn | LogicComposedSnapshotIn[]
): ILogicComposedStoreSnapshotIn {
  return (Array.isArray(snapshots) ? snapshots : [snapshots]).reduce(
    (a, snapshot) => {
      const { _id, info, composition } = snapshot
      a.info.push({ _id, ...info })
      a.composition.push({ _id, ...composition })
      return a
    },
    { info: [] as any[], composition: [] as any[] }
  )
}

export function createLogicNativeStoreSnapshot(
  snapshots: LogicNativeSnapshotIn | LogicNativeSnapshotIn[]
): ILogicComposedStoreSnapshotIn {
  return (Array.isArray(snapshots) ? snapshots : [snapshots]).reduce(
    (a, snapshot) => {
      const { _id, info } = snapshot
      a.info.push({ _id, ...info })
      return a
    },
    { info: [] as any[] }
  )
}

export function addIdsToSnapshot(snapshot: LogicComposedSnapshotIn) {
  const { _id, info, composition } = snapshot
  return {
    _id,
    info: { _id, ...info },
    composition: { _id, ...composition },
  }
}
