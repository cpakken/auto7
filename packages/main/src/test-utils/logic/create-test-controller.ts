import { objectToArray } from "@utils/collection-fns"
import { LogicController, ILogicController, LogicNativeSnapshotIn, LogicComposedSnapshotIn } from "@main/controllers/logic"
import { createLogicNativeStoreSnapshot, createLogicComposedStoreSnapshot } from "./snapshot-factories"

import * as nativeSnapshots from "./native/snapshots"
import * as composedSnapshots from "./composed"

export function createTestLogicController(env?): ILogicController {
  const nativeLoaders: LogicNativeSnapshotIn[] = objectToArray(nativeSnapshots)
  const native = createLogicNativeStoreSnapshot(nativeLoaders)

  const composedLoaders: LogicComposedSnapshotIn[] = objectToArray(composedSnapshots)
  const composed = createLogicComposedStoreSnapshot(composedLoaders)

  const snapshot = { composed, native }

  //TODO process cachedKeys
  return LogicController.create(snapshot, env)
}
