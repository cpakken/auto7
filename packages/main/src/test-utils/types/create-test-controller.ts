import { objectToArray } from "@utils/collection-fns"
import { ITypeController, TypeController, TypeNativeSnapshotIn } from "@main/controllers/types"
import { createTypeNativeStoreSnapshot } from "./snapshot-factories"

import * as nativeSnapshots from "./native/std.config"

export function createTestTypeController(env?): ITypeController {
  const nativeLoaders: TypeNativeSnapshotIn[] = objectToArray(nativeSnapshots)
  const native = createTypeNativeStoreSnapshot(nativeLoaders)

  // const composedLoaders: LogicComposedSnapshotIn[] = objectToArray(composedSnapshots)
  // const composed = createLogicComposedStoreSnapshot(composedLoaders)

  const snapshot = {
    native,
    // composed,
  }

  //TODO process cachedKeys
  return TypeController.create(snapshot, env)
}
