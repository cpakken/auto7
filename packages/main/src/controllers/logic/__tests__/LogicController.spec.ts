import { createTestLogicController } from "@main/test-utils/logic/create-test-controller"
import { when, autorun, _resetGlobalState } from "mobx"
import { ILogicController } from ".."

describe("LogicController", () => {
  let logicController: ILogicController
  beforeEach(() => (logicController = createTestLogicController()))
  afterEach(_resetGlobalState)

  test("load compound example", async () => {
    const compound = logicController.composed.get("ADDSTRING_COMPOUND")
    expect(compound.shallowInfoReady).toBe(true)
  })

  test("logicNative module loading", async () => {
    const addNative = logicController.native.get("_std_ADDSTRING")

    expect(addNative.isLoaded).toBe(false)

    await when(() => !!addNative.module)
    expect(addNative.module).toMatchSnapshot()
    expect(addNative.isLoaded).toBe(true)
  })

  test("isLoaded", async () => {
    const compound = logicController.composed.get("ADDSTRING_COMPOUND")
    const disposer = autorun(() => compound.isLoaded)
    expect(compound.isLoaded).toBe(false)
    await when(() => compound.isLoaded)
    expect(compound.isLoaded).toBe(true)
    disposer()
  })

  test("getUpdatedLogicKeys", async () => {
    const compound = logicController.composed.get("ADDSTRING_COMPOUND")
    const keys = await compound.getKeysWhenReady()
    expect(keys).toMatchSnapshot()
    expect([...keys]).toEqual(compound.info!.keysCache)
  })
})
