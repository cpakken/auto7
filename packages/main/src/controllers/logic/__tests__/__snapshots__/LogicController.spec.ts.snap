// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogicController getUpdatedLogicKeys 1`] = `
Set {
  "ADDSTRING_COMPOUND",
  "ADDSTRING",
  "_std_ADDSTRING",
  "ADDSTRING_NESTED",
}
`;

exports[`LogicController logicNative module loading 1`] = `
Object {
  "ins": Array [
    "in1",
    "in2",
  ],
  "outs": Object {
    "out": [Function],
  },
}
`;
